### ISSUE TITLE
Script Loading: Remove support for delayed inline scripts

### ISSUE THREAD

**Issue by joemcgill** [10up/wordpress-develop#69]
## Background
Our current implementation of [Enhancing the WP_Scripts API With a Loading Strategy](https://github.com/WordPress/wordpress-develop/pull/4391) includes logic to keep full backward-compatible support for inline scripts attached to registered script handles. This ensures we retain execution order of the inline scripts relative to their dependencies, even when a script that was previously registered without an `async` or `defer` attribute is updated to have one applied. It does so by delaying the execution of inline scripts until after any scripts that they depend on have finished loading by initially printing these scripts without a `text/javascript` type and then replacing them on an event listener attached to load events. This has been implemented in a way that is fully [CSP Strict compatible](`https://csp.withgoogle.com/docs/strict-csp.html`).

Even so, this aspect of the proposal continues to be questioned and considered "not a good idea". [See related Trac discussion](https://core.trac.wordpress.org/ticket/12009#comment:113). Continued debate about this aspect of the feature could put the entire feature at risk.

## Objective
In order to unblock the main objective of adding support for `async` and `defer` strategies to the `WP_Scripts` API while the implementation for inline scripts is being debated, we should create a new PR that removes the logic that delays inline scripts entirely from our current PR.

## Requirements
* All inline scripts added in the `before` or `after` position should printed with `type="text/javascript"` (same a current behavior in WP 6.2)
* All inline scripts added in the `after` position should cause the handle they're added to (and all its dependents) to be printed without a loading strategy (i.e. a blocking script).
* An inline scripts adding in the `before` position should have no effect on the eligible loading strategies of scripts it depends on.
* All Unit tests for inline scripts should be updated to confirm this behavior, or removed if unneeded.

## Implementation details
* Logic should likely be added to `has_only_deferrable_dependents` to check for the existence of inline `after` scripts while determining if dependents are deferrable so that we avoid recursively checking dependencies more than once.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by joemcgill** [comment]
This is currently blocked by https://github.com/10up/wordpress-develop/pull/67

**Comment by 10upsimon** [comment]
Based on internal comms, due to `has_non_standalone_inline_script` being removed in  https://github.com/10up/wordpress-develop/pull/67 we will need to replace current logic in `has_only_deferrable_dependents` that calls this method, currently this:

```php
if ( $this->has_non_standalone_inline_script( $dependent, 'before' ) ) {
    return false;
}
```

To something like this:

```php
if ( ! empty( $this->get_data( $handle, 'after' ) ) ) {
    return false;
}
```

This in turn will ensure that dependents with `after` handle scripts are not marked as truthy for `$this->has_only_deferrable_dependents( $handle )` checks, and this can be used within `get_eligible_loading_strategy` to ensure a blocking strategy is used.

**Comment by westonruter** [comment]
If we remove inline script handling from the core merge for the next release, it would be great to try to see if we can include that as a feature in the PerfLab plugin to test.

**Comment by westonruter** [comment]
I did a survey of `after` script usage in plugins, themes, and core.

## Plugins

There are [5,388 instances](https://wpdirectory.net/search/01H2EJ1E82DX0JCAMYSS35GD0T) of adding inline scripts in any position, and [358 instances](https://wpdirectory.net/search/01H2EJ3CEZMF6T6Q3M7X89PS6N) are specifically passing `before`. Since the default is `after`, this means the vast majority are `after` scripts: 5,030 instances (93%).

Caveat: There are [45,938 instances](https://wpdirectory.net/search/01H2EJB72DZBEA7HJ6XBSJG9F9) of `wp_localize_script()` being used. These are all essentially `before` inline scripts. Note that since this `extra` data is _always_ exported as a global variable, it can always safely be run synchronously and never needs to be delayed. Note also that a very great many of these usages should be converted to use `wp_add_inline_script()` since `wp_localize_script()` is intended specifically for strings and not arbitrary data.

## Themes

There are [808 instances](https://wpdirectory.net/search/01H2EM8SFQAB67C1TSRHZS2AJA) of adding inline scripts in any position, and [77 instances](https://wpdirectory.net/search/01H2EMANQHGSSVZ1WZ9DZW1H10) are specifically for `before` scripts. So this means 90% instances are `after` scripts.

There are two usages in core themes for `after` scripts:

1. [wp-content/themes/twentytwenty/functions.php#L662-L669](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-content/themes/twentytwenty/functions.php#L662-L669) 
2. [wp-content/themes/twentytwentyone/functions.php#L439-L447](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-content/themes/twentytwentyone/functions.php#L439-L447)

Caveat: There are [8,777 instances](https://wpdirectory.net/search/01H2TQ3MDH473FD3XDJF5N2FEE) of `wp_localize_script()`

## Core

There are 20+ instances of `after` scripts being used in core:

1. [wp-admin/site-editor.php#L94-L114](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/site-editor.php#L94-L114) 
1. [wp-admin/plugin-editor.php#L152-L153](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/plugin-editor.php#L152-L153) 
1. [wp-includes/widgets/class-wp-widget-media-audio.php#L159-L179](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/widgets/class-wp-widget-media-audio.php#L159-L179) 
1. [wp-includes/script-loader.php](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/script-loader.php) -- There are 15 instances in this file, and only two of them are `before` scripts.
1. [wp-admin/edit-form-blocks.php#L79-L83](https://github.com/WordPress/wordpress-develop/blob/5dcb2e928c8e9643c5a5c919f0b63df7bb17c5e3/src/wp-admin/edit-form-blocks.php#L79-L83)
1. [wp-includes/widgets/class-wp-widget-text.php#LL438C31-L438C31](https://github.com/WordPress/wordpress-develop/blob/5dcb2e928c8e9643c5a5c919f0b63df7bb17c5e3/src/wp-includes/widgets/class-wp-widget-text.php#LL438C31-L438C31) 
1. [wp-includes/block-editor.php#L642-L649](https://github.com/WordPress/wordpress-develop/blob/5dcb2e928c8e9643c5a5c919f0b63df7bb17c5e3/src/wp-includes/block-editor.php#L642-L649) 
1. [wp-includes/class-wp-customize-widgets.php#L869-L873](https://github.com/WordPress/wordpress-develop/blob/5dcb2e928c8e9643c5a5c919f0b63df7bb17c5e3/src/wp-includes/class-wp-customize-widgets.php#L869-L873) 
1. [wp-admin/includes/class-wp-privacy-policy-content.php#L345-L353](https://github.com/WordPress/wordpress-develop/blob/5dcb2e928c8e9643c5a5c919f0b63df7bb17c5e3/src/wp-admin/includes/class-wp-privacy-policy-content.php#L345-L353) 
1. [wp-includes/widgets/class-wp-widget-media-gallery.php#L147-L167](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/widgets/class-wp-widget-media-gallery.php#L147-L167) 
1. [wp-includes/widgets/class-wp-widget-custom-html.php#L217-L235](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/widgets/class-wp-widget-custom-html.php#L217-L235) 
1. [wp-includes/widgets/class-wp-widget-media-video.php#L198-L218](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/widgets/class-wp-widget-media-video.php#L198-L218) 
1. [wp-admin/includes/template.php#L2795-L2802](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/includes/template.php#L2795-L2802) 
1. [wp-includes/widgets/class-wp-widget-media-image.php#L306-L326](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/widgets/class-wp-widget-media-image.php#L306-L326) 
1. [wp-includes/class-wp-customize-widgets.php#L853-L873](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/class-wp-customize-widgets.php#L853-L873) 
1. [wp-admin/includes/post.php#L2322-L2399](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/includes/post.php#L2322-L2399) 
1. [wp-admin/edit-form-blocks.php#L79-L303](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/edit-form-blocks.php#L79-L303) -- Of four instances, only [one](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/edit-form-blocks.php#L124-L128) is a `before` script.
1. [wp-admin/theme-editor.php#L146-L147](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/theme-editor.php#L146-L147) 
1. [wp-admin/widgets-form-blocks.php#L38-L58](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-admin/widgets-form-blocks.php#L38-L58) 
1. [wp-includes/general-template.php#L3909](https://github.com/WordPress/wordpress-develop/blob/ba9a2f8b83211fbdee1621fd4f39dcb11908b817/src/wp-includes/general-template.php#L3909)

**Comment by felixarntz** [comment]
Thanks @westonruter for the detailed report! Based on that, I think we can conclude that technically `before` scripts are still _way_ more common than `after` scripts, simply due to `wp_localize_script()`. It's right that a lot of those should probably be updated to use `wp_add_inline_script()` with `before`, but I guess technically they work in a similar way. It's also worth noting that any inline script from `wp_localize_script()` can automatically be considered "standalone", i.e. not requiring any code from potential dependencies.

While looking at `wp_add_inline_script()` specifically heavily favors the use of `after` inline scripts, when putting the results together with `wp_localize_script()`, ~90% of inline scripts appear before the main script for plugins and ~92% for themes. So overall the relative impact will still be very high even if any script with an `after` inline script is force-blocking.

**Comment by felixarntz** [comment]
However, following on the above, that still leaves us with ~10% of inline scripts across all plugins and themes that make it impossible for its handle to be deferred or asynchronous. While that is not a blocker for a first implementation, we should definitely make it _possible_.

Something I'd like us to consider is, if we do not move forward with the delaying of inline scripts, is there another simpler way to at least _allow_ developers to have such inline scripts and still defer. For example, the `$standalone` approach that originally was part of #4391, would allow exactly that without any of the complicated workarounds:
* If the inline `after` scripts of a handle are all marked as "standalone", the script can just remain as is. The developer should only specify that if they have written the JS code in a way that waits for whatever it needs from the main script.
* If any inline `after` script of a handle is not marked as "standalone" (which of course will remain the default), its handle would remain force-blocking.

I would like to understand why that parameter was removed from the implementation, I don't recall or maybe I missed that. Especially if we pull out the inline script handling, I think bringing it back would be a useful idea to allow for scripts with inline `after` scripts to be deferred.

My perspective as a summary: If we cannot land the approach of automatically making inline `after` scripts work with `defer` and `async`, that's unfortunate. But if we don't provide a way to even allow developers to make it work, that's much worse. It would be an artificial limitation.

cc @joemcgill @10upsimon

**Comment by westonruter** [comment]
> I would like to understand why that parameter was removed from the implementation, I don't recall or maybe I missed that. Especially if we pull out the inline script handling, I think bringing it back would be a useful idea to allow for scripts with inline `after` scripts to be deferred.

@felixarntz Standalone scripts were removed in https://github.com/10up/wordpress-develop/pull/67 because there was no longer a need for them since both `before` and `after` scripts could be delayed automatically.

**Comment by westonruter** [comment]
> I think we can conclude that technically `before` scripts are still _way_ more common than `after` scripts, simply due to `wp_localize_script()`. It's right that a lot of those should probably be updated to use `wp_add_inline_script()` with `before`, but I guess technically they work in a similar way. It's also worth noting that any inline script from `wp_localize_script()` can automatically be considered "standalone", i.e. not requiring any code from potential dependencies.
> 
> While looking at `wp_add_inline_script()` specifically heavily favors the use of `after` inline scripts, when putting the results together with `wp_localize_script()`, ~90% of inline scripts appear before the main script for plugins and ~92% for themes. So overall the relative impact will still be very high even if any script with an `after` inline script is force-blocking.

@felixarntz While `wp_localize_script()` is _essentially_ a `before` script in that it prints before, it is technically an `extra` script that is printed before. Since `wp_localize_script()` can _only_ output a (string-ish) data structure which is populated into a global variable and cannot have any logic at all, I'd lean toward considering it not really a script but more of a data export mechanism. In this way `wp_localize_script()` is more similar to `wp_set_script_translations()` (which also prints before) than it is to `wp_add_inline_script( ..., 'before' )`.

Not sure if this factors into the weights at all, but I'll also say that just because a lot of plugins are using `wp_localize_script()` to export data before, this doesn't mean they necessarily should be printing such before scripts anymore. In particular, any time there is a before script this necessitates populating the global scope with a variable that the subsequent external script has to know about. In contrast, when looking at the non-before examples of the `wp_add_inline_script()`, the data is passed straight into the API exposed by the script without the need for any global variable. So it's much cleaner and easier to understand.

**Comment by westonruter** [comment]
> My perspective as a summary: If we cannot land the approach of automatically making inline `after` scripts work with `defer` and `async`, that's unfortunate. But if we don't provide a way to even allow developers to make it work, that's much worse. It would be an artificial limitation.

@felixarntz If an `after` inline script is going to have the effect of forcing a delayed script to be blocking, rather than resorting to re-adding the standalone concept, I think it would be better to just require developers to instead use a `before` inline script that has a `DOMContentLoaded` event handler.

## Ideal

```php
wp_enqueue_script( 'foo', '/foo.js', array(), '0.1', [ 'strategy' => 'defer' ] );
wp_add_inline_script(
    'foo',
    sprintf(
        'Foo.init(%s);', 
        wp_json_encode( $data ) 
    ),
    'after' 
);
```

## Compromise

```php
wp_enqueue_script( 'foo', '/foo.js', array(), '0.1', [ 'strategy' => 'defer' ] );
wp_add_inline_script(
    'foo',
    sprintf( 
        'document.addEventListener( "DOMContentLoaded", () => { Foo.init(%s); } );',
        wp_json_encode( $data ) 
    ), 
    'before' 
);
```

**Comment by felixarntz** [comment]
@westonruter 
> Standalone scripts were removed in https://github.com/10up/wordpress-develop/pull/67 because there was no longer a need for them since both `before` and `after` scripts could be delayed automatically.

Thanks! Makes sense from that perspective. Depending on the outcome of our conversation here, we could consider bringing it back though if we remove that inline script handling.

> While `wp_localize_script()` is _essentially_ a `before` script in that it prints before, it is technically an `extra` script that is printed before. Since `wp_localize_script()` can _only_ output a (string-ish) data structure which is populated into a global variable and cannot have any logic at all, I'd lean toward considering it not really a script but more of a data export mechanism.

I understand that they technically work in a slightly different way, but even if they were using `wp_add_inline_script()` with `before`, it would still be a data export mechanism. I agree with you that populating data into an API is a cleaner solution, but communicating that to the majority of plugin developers out there is probably not feasible. Given `wp_localize_script()` is so heavily used, I think it's safe to say that data export is by far the most common use-case for inline scripts printed before the main script.

> If an `after` inline script is going to have the effect of forcing a delayed script to be blocking, rather than resorting to re-adding the standalone concept, I think it would be better to just require developers to instead use a `before` inline script that has a `DOMContentLoaded` event handler.

That could work, but it means that whatever that inline script does will only run on `DOMContentLoaded`, i.e. later than potential dependents of the main script. Of course it's just a compromise as you say, but I think a better compromise would be to listen to the main script element's `load` event itself, which for a simple solution requires an `after` inline script. Being able to declare such a script as "standalone" would solve the problem in a cleaner way. (That said, maybe "standalone" is not the good name for that parameter anymore, but you get the gist - it's more of a "considers that the script is deferred/async, so please don't force it to blocking" parameter).

Basically:

```php
wp_enqueue_script( 'foo', '/foo.js', array(), '0.1', [ 'strategy' => 'defer' ] );
wp_add_inline_script(
    'foo',
    sprintf(
        'document.querySelector( "#foo-js" ).addEventListener( "load", () => { Foo.init(%s); } );',
        wp_json_encode( $data )
    ),
    'after',
    true // Assuming this is the "standalone" parameter.
);
```

**Comment by joemcgill** [comment]
Thanks both. It seems like we're mostly aligned with the second requirement listed above:

> * All inline scripts adding in the after position should cause the handle they're added to (and all its dependents) to be printed without a loading strategy

However, it seems like @felixarntz your proposing we add another requirement that would allow an inline script to be registered with an `after` position in a way that would not cause the handle to become blocking. Given that anything that we implement to meet that requirement would have to be opt-in anyway, I'd prefer our initial implementation not change the function signature of `wp_add_inline_script` and instead start with documentation for how to work around this limitation (e.g., @westonruter's proposed compromise above).

There are two use cases that we need to keep in mind here:

1. Both the main handle and the inline script are being registered by the same plugin or theme.
2. The main handle is registered by core, a plugin, or theme and a separate plugin or theme adds the inline script.

For the former, I think it's reasonable to expect that if the developer wanted to add a delayed loading strategy to the main handle, then they would also modify the inline script to get the desired result.

For the latter, there is no way for the original developer to ensure third-party code isn't going to effect the intended loading strategy, even if we do update the function signature for `add_inline_scripts`. The only alternative I can think of for this case is to provide a filter  to `WP_Scripts::get_eligible_loading_strategy` (or similar) so the fallback strategy can be "fixed" if a third-party inline script causes the intended strategy to be reset to "blocking".

No matter what workarounds we use, the only way I see to fully support inline scripts on delayed script handles AND maintain execution order that is backwards compatible is to keep the current (or similar) implementation that we have now, so I don't think we can attempt to remove support for delayed inline scripts while ALSO trying to maintain strict execution order.

If the way we've already solved this problem becomes a blocker for merge, I'd much rather have the initial implementation that we know has some limitations and focus on addressing the blocking concerns rather than introduce short-term API changes that make future-compat more difficult.

**Comment by westonruter** [comment]
> Of course it's just a compromise as you say, but I think a better compromise would be to listen to the main script element's `load` event itself, which for a simple solution requires an `after` inline script. 

@felixarntz That's true. An `after` inline script for a `defer` script could attach a `load` event directly onto the `script` element. However, this does not work for `async` scripts since they may have already loaded. (Granted, by being async they should be inherently more flexible with load order.) The various `async` methods to do the same are explored in [async-script-load-listeners.glitch.me](https://async-script-load-listeners.glitch.me/), the last of which was implemented in #67 via a delegated capturing `load` event handler. This same approach would have to be copied for any such script adding an inline `after` script for `async` scripts. So the nice thing about the PR now is that nobody needs to reinvent the wheel as it is all handled automatically.

**Comment by felixarntz** [comment]
@joemcgill 
> There are two use cases that we need to keep in mind here:
>
> 1. Both the main handle and the inline script are being registered by the same plugin or theme.
> 2. The main handle is registered by core, a plugin, or theme and a separate plugin or theme adds the inline script.
>
> For the former, I think it's reasonable to expect that if the developer wanted to add a delayed loading strategy to the main handle, then they would also modify the inline script to get the desired result.

What I'm trying to convey is that they won't be able to do that if we remove delayed inline script support without bringing some version of the "standalone" approach back. They wouldn't be able to use an `after` script, so they would be forced to use a `before` script in which then they would need to rely on `DOMContentLoaded`, which can be problematic for dependents. So this more a problem about your point 1 than point 2.

> For the latter, there is no way for the original developer to ensure third-party code isn't going to effect the intended loading strategy, even if we do update the function signature for `add_inline_scripts`.

That is never the case I think, but I don't think that was ever a goal either. Even in the current core PR, another plugin developer can e.g. enqueue a blocking script that depends on a deferred script from another plugin and in that way change the loading strategy of the other plugin's script.

> No matter what workarounds we use, the only way I see to fully support inline scripts on delayed script handles AND maintain execution order that is backwards compatible is to keep the current (or similar) implementation that we have now, so I don't think we can attempt to remove support for delayed inline scripts while ALSO trying to maintain strict execution order.

I think that's only partially correct, particularly for the edge-case that a `before` script depends on a previous dependency's code. For an `after` script however, I think my suggested approach of a "standalone" parameter (for the lack of a better name for now) will still allow us to maintain strict execution order just as well as it worked _prior_ to implementing #67 (which also added that for `before` scripts). The only difference between that and the core PR without #67 would be that our current implementation delays `after` inline scripts by default, while the "standalone" driven approach would be opt-in (i.e. the developer would implement their inline script in a way that delays execution themselves and then mark their inline script to let WP core know that it should not make the main script force-blocking).

**Comment by felixarntz** [comment]
Just to clarify my points here:
* I would prefer to land the current implementation of the core PR.
* If that is not feasible due to the pushback, then we remove the delayed script support entirely.
* The one thing that I think is a critical limitation of that though is that it will be entirely impossible to have a `defer` or `async` dependency with an inline `after` script. Having a way for a developer to opt in to achieve that IMO is crucial - basically some flag to indicate to WP core "please don't change the loading strategy of my defer/async script even though you see that there is an `after` script".
* We could certainly implement the point above in a separate follow up PR that goes into another commit after the initial one. At that point, we would have to evaluate the tradeoffs and feedback on whether we go with the approach of "core solves it automatically" (#67) or "plugin developer opts in" (see above).

**Comment by joemcgill** [comment]
I agree with you on points 1 and 2 above. This ticket is specifically to address point two.

> The one thing that I think is a critical limitation of that though is that it will be entirely impossible to have a `defer` or `async` dependency with an inline after script

I definitely agree with your description here. This is largely the point of this ticket in the first place, as some of the pushback as been that the Script Loader should NOT support after inline scripts on delayed scripts.

> We could certainly implement the point above in a separate follow up PR

In that case, I'd propose that we keep the scope of this ticket to the removal of support and create a new ticket to address the limitation that this would introduce. My hunch is that following that path will lead us back  to the original implementation in the first place.

**Comment by felixarntz** [comment]
> some of the pushback as been that the Script Loader should NOT support after inline scripts on delayed scripts

Is the pushback really that this should not be supported at all? The way I understand the pushback, it is about the way this would need to be implemented in core to happen automatically, not about that it generally should not be possible.

> In that case, I'd propose that we keep the scope of this ticket to the removal of support and create a new ticket to address the limitation that this would introduce.

Sounds good to me.

> My hunch is that following that path will lead us back to the original implementation in the first place.

Do you mean because we would be deciding that the original is the better choice? I think the two approaches are still distinctly different. The "standalone" approach is extremely simple from a core perspective, but puts a bit more burden on the plugin developer. The current / original approach is far more complicated from a core perspective, but makes things very easy for the plugin developer.

I hope it's clear what I mean by the "standalone" approach, it wouldn't entail any of the inline script JS / `text/template` code that is currently part of the implementation.

**Comment by joemcgill** [comment]
> Is the pushback really that this should not be supported at all?

I've understood the feedback in https://core.trac.wordpress.org/ticket/12009#comment:115 to be both specific to our implementation but also a general opinion that supporting inline scripts printed after a handle is a bad idea in the first place. For example:

> "There is actually very little need for "after" scripts, so this feature of Script Loader has always been on the border of being quite useless. I.e. nearly all cases where an "after" script is used can be handled by using a "before" script. There are only few rare exceptions (which frankly do not warrant the existence of the convenience method for adding of "after" scripts, but that's a different question of a "bad API decision" that was made some time ago)."

I do understand the difference between what you are proposing and what our current implementation achieves and I do think we should consider that approach as an enhancement if we aren't able to fully support inline scripts. Either way, it sounds like we're in agreement that doing so is out of scope of this specific ticket and we'll pick that up in a separate one.

**Comment by joemcgill** [comment]
For transparency: I've also raised this question in today's WP Core chat and asked for feedback in the [#core-js channel on Slack](https://wordpress.slack.com/archives/C5UNMSU4R/p1686774196700899).

**Comment by felixarntz** [comment]
Thanks for clarifying @joemcgill. I disagree with the following bit of the comment you referenced though:
> nearly all cases where an "after" script is used can be handled by using a "before" script

When the inline script relies on e.g. a function that the main script defines, how would that work in a `before` script? Sure, we can wrap things in a `DOMContentLoaded`, but that may not always be feasible since that means the script will delay execution until that event. An `after` script today has the purpose to run code right after the main script is loaded, so changing to use a `before` script with `DOMContentLoaded` can break expectations in terms of interoperability.

The example I mentioned in https://github.com/10up/wordpress-develop/issues/69#issuecomment-1589894334 is something which I believe is not reasonably possible in a `before` script. And the fact that 90% of `wp_add_inline_script()` instances are `after` scripts shows that `after` scripts in general are certainly not an edge-case.

I understand the pushback against the complex implementation, but I disagree that `after` inline scripts as a concept are useless.
